### ISSUE TITLE
finegrained-fp8: grouped matmul writes past N and reads past N/K (wrong results on ordinary expert shapes)

### ISSUE THREAD

**Issue by malaiwah** [huggingface/kernels-community#1153]
## Summary

`matmul_grouped` returns wrong results for a wide range of ordinary expert shapes. The grouped kernels bound their tiles in the M direction only: `store_tile` masks the store with `row_mask` but never with `offs_bn < N`, and the tensor-scale and MX kernels load `b` (and `a`) with no K bound. Whenever a tile extends past `N` or `K`, the kernel reads outside the weight tensor and, because `C` is a contiguous `(S, N)` tensor with `stride_cm == N`, writes the overhanging columns into the *next rows* of the output.

The N overhang is not just an unaligned-shape problem: the autotuner is free to pick `BLOCK_SIZE_N` larger than `N`, so even `N = 128` produces a partial tile with `BLOCK_SIZE_N = 256`. That is why the failures below appear on power-of-two widths too, and why they are intermittent — whether the correct value or the overhang from the previous row lands last in a contended column is unspecified.

The block-scale route is safe: `_w8a8_block_dynamic_fp8_matmul_grouped` asserts `N % block_n == 0` and `K % block_k == 0` and pins `BLOCK_SIZE_N`/`BLOCK_SIZE_K` to those values. The tensor-scale and MX routes have no such assert and take their block sizes from autotune.

## Where

On `main` (`finegrained-fp8/torch-ext/finegrained_fp8/grouped.py`, commit `92b1518`):

- `store_tile` line 138: `tl.store(c_ptrs, c, mask=row_mask[:, None])` — no `offs_bn < N` term, and `offs_bn` comes from `pid_n * BLOCK_SIZE_N + tl.arange(0, BLOCK_SIZE_N)` (line 126) with grid `triton.cdiv(N, META["BLOCK_SIZE_N"])` (lines 559, 684). Shared by all three grouped kernels.
- `w8a8_tensor_dynamic_fp8_matmul_grouped_kernel` lines 302-303: `a`/`b` loaded with no K mask, `b` with no N mask.
- `mxfp_dynamic_matmul_grouped_kernel` lines 408-412: same, plus the unmasked `Bs` load.

## Reproducer

No model, no `transformers`, CUDA only. Every row is a constant power of two, so the FP8/MX quantization of each row is exact and `C[i, j]` must be exactly `K * 2**i` — any deviation is the kernel leaving its logical extent, not a rounding artifact.

```python
import torch
from kernels import get_kernel

kernel = get_kernel("kernels-community/finegrained-fp8", version=4)
E, ROWS, N, K = 2, 2, 33, 64          # N deliberately not a multiple of BLOCK_SIZE_N
S = E * ROWS
A = (2.0 ** torch.arange(S, dtype=torch.float32)).unsqueeze(1).expand(S, K)
A = A.contiguous().to("cuda", torch.bfloat16)
counts = torch.full((E,), ROWS, device="cuda", dtype=torch.int32)
offsets = counts.cumsum(0).to(torch.int32)

B = torch.ones((E, N, K), device="cuda").to(torch.float8_e4m3fn)          # MXFP8 weights
Bs = torch.ones((E, N, K // 32), device="cuda").to(torch.float8_e8m0fnu)  # UE8M0 scales

C = kernel.matmul_grouped(A, B, Bs, offsets, counts, None)
expected = (A[:, :1].float() * K).expand(S, N)
print(C.float()[:, 0].tolist())        # [64.0, 128.0, 256.0, 0.0]  <- last row clobbered
print(expected[:, 0].tolist())         # [64.0, 128.0, 256.0, 512.0]
torch.testing.assert_close(C.float(), expected, atol=0, rtol=0)
```

Observed on that shape: 31 of 132 elements wrong, all of them row 3 columns 0-30 — exactly the 31-column overhang of row 2's tile (`N = 33`, `BLOCK_SIZE_N = 64`), carrying values accumulated from weight columns that do not exist.

## Scope

Sweep of the same reproducer, 3 repeats per shape, `matmul_grouped` with `output_dtype=None`:

| route | N | K | repeats wrong (stock) | repeats wrong (patched) |
|---|---|---|---|---|
| tensor | 32 | 64 | 0/3 | 0/3 |
| tensor | 33 | 64 | 2/3 | 0/3 |
| tensor | 64 | 64 | 1/3 | 0/3 |
| tensor | 65 | 64 | 3/3 | 0/3 |
| tensor | 128 | 64 | 0/3 | 0/3 |
| tensor | 129 | 64 | 1/3 | 0/3 |
| tensor | 128 | 96 | 3/3 | 0/3 |
| tensor | 128 | 160 | 3/3 | 0/3 |
| mxfp8 | 32 | 64 | 1/3 | 0/3 |
| mxfp8 | 33 | 64 | 1/3 | 0/3 |
| mxfp8 | 64 | 64 | 1/3 | 0/3 |
| mxfp8 | 65 | 64 | 2/3 | 0/3 |
| mxfp8 | 128 | 64 | 1/3 | 0/3 |
| mxfp8 | 129 | 64 | 0/3 | 0/3 |
| mxfp8 | 128 | 160 | 3/3 | 0/3 |
| mxfp4 | 32 | 64 | 1/3 | 0/3 |
| mxfp4 | 33 | 64 | 1/3 | 0/3 |
| mxfp4 | 65 | 64 | 3/3 | 0/3 |
| mxfp4 | 129 | 64 | 1/3 | 0/3 |
| mxfp4 | 128 | 160 | 3/3 | 0/3 |

17 of 24 swept configurations produced wrong output at least once; up to 510 wrong elements in a single call. `K = 96` and `K = 160` fail deterministically on the tensor route (K tail, `BLOCK_SIZE_K` 128); the N-overhang cases are intermittent, as expected from a contended store.

## Environment

- NVIDIA RTX PRO 6000 Blackwell Server Edition, compute capability 12.0, driver CUDA 13.0
- torch 2.14.0+cu130, triton 3.8.0, `kernels` snapshot `8c178950e8710e26a2210e4e909cd02dc16c8715` (`version=4`, `build/torch-cuda`)

`compute-sanitizer --tool memcheck` reports "Device not supported" on this GPU, so I have no sanitizer trace to attach; the evidence above is functional. On a supported device I would expect invalid `__global__` write/read reports in `store_tile` and the `b` loads.

## Fix

What worked here, on top of the `version=4` build:

1. `store_tile`: take `N` and store with `mask=row_mask[:, None] & (offs_bn[None, :] < N)`; pass `N` from all three grouped kernels.
2. Tensor-scale kernel: mask the loop loads with the K tail, `k_mask = offs_k + k * BLOCK_SIZE_K < K`, and add `offs_bn < N` to the `b` load.
3. MX kernel: same K mask for `a`, and bound `b` by clamping its indices rather than masking with `other=` — a single typed fill value cannot serve both packed-uint8 (MXFP4) and fp8 (MXFP8) weights (`other=0` raises `cannot cast int32 to fp8e4nv`). Since `a` is already zero on the K tail and the store drops the N overhang, the clamped reads cannot contribute:

```python
    b_cols = tl.minimum(offs_bn, N - 1)
    b_base = B + expert_id * stride_be + b_cols[None, :] * stride_bn
    ...
        kb = tl.minimum(
            offs_kb + k * (BLOCK_SIZE_K // VALUES_PER_BYTE), K // VALUES_PER_BYTE - 1
        )
        b = tl.load(b_base + kb[:, None] * stride_bk)
```
   and mask the `Bs` load with `(offs_bn[:, None] < N) & sf_mask[None, :]`, `other=127` (UE8M0 1.0).

With those changes all 24 swept configurations are exact (right-hand column above).

Alternatively, if unaligned expert widths are considered out of contract for these routes, the wrappers should assert `N`/`K` alignment the way the block-scale wrapper already does — but that would still leave the `BLOCK_SIZE_N > N` case, which needs the store mask regardless.

Happy to send this as a PR if you prefer.

**Comment by malaiwah** [comment]
Follow-up: the defect is independent of the FP8 hardware path, which may make it easier to reproduce on whatever card you have.

The published kernels cannot run on pre-Ada GPUs at all — on a Tesla T4 (SM 7.5, torch 2.14.0+cu130, triton 3.8.0) `matmul_grouped` dies in `fp8_act_quant` before any matmul:

```
ValueError("type fp8e4nv not supported in this architecture. The supported fp8 dtypes are ('fp8e4b15', 'fp8e5')")
```

So I re-ran only the tile mapping and epilogue — `_grouped_tile_setup`'s binary search, the same `offs_bn = pid_n * BLOCK_SIZE_N + tl.arange(0, BLOCK_SIZE_N)`, grid `cdiv(N, BLOCK_SIZE_N)`, and `store_tile` — in fp16, changing nothing but the store mask. `N = 33`, `K = 64`, `BLOCK_SIZE_N = 64`, 4 rows across 2 experts, rows holding `1, 2, 4, 8` so every `C[i, j]` must be exactly `K * 2**i`:

| store mask | runs wrong (20 runs) | wrong elements per run |
|---|---|---|
| `row_mask[:, None]` (current) | 18/20 | 31 or 62 |
| `row_mask[:, None] & (offs_bn[None, :] < N)` | 0/20 | 0 |

A failing run returns `[64.0, 128.0, 128.0, 252.0]` in column 0 instead of `[64.0, 128.0, 256.0, 512.0]`: row 1's tile overhang lands in row 2 (`128.0` where `256.0` belongs), and row 3 picks up a partial `252.0`. The first two runs of each batch are clean, which is why this hides easily in a single-shot test — the overhang and the legitimate write target the same addresses, so it depends on which store retires last.

That also means an fp16/bf16 GPU is enough to regression-test the masking, independent of FP8 support. There is no CPU path to test against (the repo ships `torch-cuda`, `torch-rocm` and `torch-xpu` builds only).

**Comment by malaiwah** [comment]
Sent the fix as #1154: the three tail-masking hunks plus a regression that drives `matmul_grouped` over the three routes and six `(N, K)` shapes (13 failed / 5 passed before, 18 passed after, SM120).
